PFW-1028: Fix recovery from filament sensor checks / crash detect / power panic

wavexx opened the pull request:
This PR fixes several issues with mesh_plan_buffer_line which cause undefined/broken behavior when the printer performs a filament check, detects a crash and/or recovers from a power panic.

The first issue is that restore_print_from_ram_and_continue can be called from a nested manage_inactivity() call. We need to properly abort the topmost process_commands() loop in order to avoid smashing the stack and essentially putting the printer in a random state.

A common scenario is a stalled plan_buffer_line which recovers from:

- a filament check
- a paused print
- a stopped print
- a crash detect

We then abort the state correctly also in the mesh_plan_buffer_line call, first by not corrupting the current position of the head, and also by aborting early if we discover we're in an aborted recursive call of process_commands().

Although this "looks" uncommon, a frequent scenario is:

0 process_commands
1 mesh_plan_buffer_line
2 plan_buffer_line
3 filament_check => manage_inactivity
4 process_commands

This PR is based on top on the previous PR #2271 in order to perform proper command buffer flushing with a starved command buffer in order to work at all.

In order to recover properly with an immediate relative move after panic we need to store the amount of relative movement which has been done for the current gcode instruction. This is not possible to perform for a chunked move, such as the one produced by mesh_plan_buffer_line (since the move might still be partially planned *while* the panic occurs). As such, we store the calculated absolute target position for each plan entry instead, which allows us to calculate the relative movement required easily from the current position.

wavexx committed:
Clear "sdprinting" state only when all SD moves are complete

Ensure card.printingHasFinished sees all the planned moves before
clearing sdprinting. To do that, we need to ensure all SD commands
exited the command queue.

wavexx committed:
Unroll the stack in restore_print_from_ram_and_continue

While handling moves in a recursive plan, such a filament check,
ensure restore_print_from_ram_and_continue unwinds the stack by
aborting early from any call that waits on the planner.

This currently only handles G1 moves, but hard-coded behavior that can
trigger recursive behavior (such as filament change) will probably have
to be checked too.

wavexx committed:
Do not set current_position in mesh_plan_buffer_line

This is a task which _has_ to be performed by the caller,
such as prepare_move().

wavexx committed:
Replace a few loops with set_current_to_destination()

wavexx committed:
Fix recovery from relative/chunked moves

When starting to replay existing USB/SD commands from a recovery state,
an immediate relative move needs to compensate for a previously
interrupted move. This is almost the norm for the E axis.

Instead of saving the relative status of the move (which needs to
account for the world2machine conversion and is not always available on
a chunked move split by MBL) save directly the calculated target
position for the move in the original plan, which is easy to replay.

leptun commented:
@wavexx Do these changes also fix mesh_plan_buffer_line creating stuttering when starting from an empty buffer?

wavexx commented:
I was hoping it would fix that, but no :(.

There's some extra fixing required on the planner/stepper which conflicts with LA1.5, so I decided to stop there as it would get messy later otherwise.

wavexx committed:
Fix M600 insertion by the filament sensor

Remove incorrect usage of stop_and_save_print combined with the
fsensor_recovert internal instruction which would result in a
broken sequence of events and/or broken stack.

Re-use the now safe stop/recover functions in the same spot
(fsensor_checkpoint_stream) to effectively cut a hole in the current
gcode stream to insert an M600 instruction, which removes all
recursive behavior without the need of extra state variables.

wavexx committed:
Save/restore feedmultiply independently of the feedrate

wavexx committed:
Save/restore the default feedrate correctly

Since the global feedrate can be similarly modified for moves ahead of
time, save the original feedrate in the planner as we do for
gcode_target.

This avoids having to undo feedmultiply (and machine limits!) from
"nominal_speed" as previously done.

Thanks @leptun

wavexx committed:
Halve memory usage of saved feedrate

Truncate the saved feedrate to an uint16_t.
This is more than sufficient for recovery.

wavexx committed:
Fix pause/resume when using M25/M601

Remove the conflicting and mostly useless card.paused flag (the printing
is either paused, or not) and switch to isPrintPaused only which
accounts for both cases (SD/USB) correctly.

Fix M27/getStatus to show the current real status of the SD print.

Synchronize the queue on M601, as required to precisely pause the print
at the correct instruction.

Alias M25 to M601, which when combined with PR #1899 fixes issue #1614.

Guard against incorrect usage in M601, M602 and M603.

wavexx commented:
This became a big melting-pot of several pause/resume state fixes, which was not my original plan, but they're all tightly interconnected.
